### ISSUE TITLE
EXP-001 research: frozen M2BERT symbolic representation probe

### ISSUE THREAD

**Issue by gr-rr** [ListenCloser/listencloser#1304]
Program owner: #1303.
Research synthesis: #1297.
Canonical landscape: #1246.
Relevant current product owners: #1233 / PR #1267 (Melody), #1300 (bespoke MIR cleanup), #1178 (symbolic detail).

## Question

> **Can a modern pretrained symbolic representation make multiple ListenCloser note-level tasks linearly/simple-MLP separable, without training the backbone?**

This is the first active low-cost ML pilot. It is deliberately a **representation probe**, not a production engine integration or a new foundation-model program.

## Why this is worth testing

Current ListenCloser symbolic logic spans several separate mechanisms: MidiBERT/LStoM Melody selection, literal symbolic detail, motif controls, theory candidates, and bespoke melody-derived thresholds. M2BERT (2025) reports gains over MidiBERT across many symbolic tasks using music-aware pretraining objectives rather than generic BERT masking.

The a-priori hypothesis is:

```text
exact symbolic notes
→ frozen pretrained contextual representation
→ tiny task-specific head
```

may replace several one-off heuristic features without requiring large-data training.

## Upstream starting point

Primary candidate:
- M2BERT paper: https://arxiv.org/abs/2507.04776
- official implementation: https://github.com/york135/M2BERT
- associated SMCBenchmark: https://github.com/york135/SMCBenchmark

Do not assume code/checkpoint commercial rights merely because assets are public. Record exact repository/checkpoint/data terms before any product recommendation.

## Scope

Use the pretrained M2BERT backbone **frozen**.

Extract per-note embeddings and test at least two, ideally three, bounded tasks that map to existing LC/research objects and available labels.

Preferred order:

1. **Melody membership** — note-level melody/not-melody classification; compare against the relevant current/held-out LC control and/or the same benchmark split used by the upstream family.
2. **Motif-note importance / motif membership** — compare against the current motif control where semantics align, or use a published motif-note dataset/split.
3. One additional symbolic task with clean labels and a plausible LC consumer, e.g. local-key-associated note context, texture/voice-related classification, or another SMCBenchmark task.

Do not invent a task merely to produce a positive result.

## Models / ablations

For each task, compare:

```text
A. existing/current control or task-native simple baseline
B. frozen M2BERT embedding + linear/logistic head
C. frozen M2BERT embedding + tiny 1–2 layer MLP
```

Optional only if cheap and useful:
- layer-selection probe across a small fixed subset of M2BERT layers;
- frozen MidiBERT representation as an additional comparator when the exact checkpoint/input contract is already available.

No full-backbone fine-tuning in this issue.

## Data contract

Before a result-bearing run:

- choose exact public/right-safe dataset(s);
- freeze train/dev/test split or upstream official split;
- record dataset revision/hash where practical;
- make note-identity/label mapping deterministic;
- keep training-overlap caveats explicit;
- do not use ListenCloser production/test fixtures as fake benchmark ground truth unless they have legitimate labels.

## Primary metrics

Use task-native held-out metrics:

- binary/multiclass F1 where appropriate;
- accuracy only where that is the standard task metric;
- macro and per-class diagnostics for imbalanced tasks.

The primary decision is not "does it train?" but:

> **Does frozen M2BERT materially improve held-out quality over the existing/simple baseline on at least two LC-relevant tasks?**

## Compute budget

Target:

```text
GPU-hours <= 4 total
cash cost <= $5
3 random seeds for tiny heads where training variance matters
```

Preferred execution ladder:
1. local CPU/MPS for preprocessing and, if practical, embedding extraction;
2. free Kaggle GPU for CUDA-only dependency/runtime friction;
3. bounded Modal/RunPod run only if needed.

Do not launch broad hyperparameter sweeps. Use one sensible linear head and one small MLP configuration, with only minimal predeclared variation.

## Suggested stack

- uv
- PyTorch
- scikit-learn
- upstream M2BERT code/checkpoint
- pandas/NumPy for fixed manifests/results
- task-native metrics / sklearn metrics
- optional local MLflow or plain JSON result records

No production dependency changes are required.

## Deliverables

1. Exact environment / upstream revision / checkpoint identity.
2. Fixed dataset/split manifest(s).
3. Reproducible embedding extraction command/script.
4. Result table for baseline vs linear vs tiny MLP.
5. Per-task error observations, not only macro scores.
6. Runtime/RAM/GPU observations sufficient to judge whether future research is cheap.
7. Rights note separating code, checkpoint, and data terms.
8. One explicit disposition:

```text
FAIL — representation did not beat controls meaningfully
NARROW — useful for one task only; no shared-backbone thesis yet
CONTINUE — useful on >=2 relevant tasks; graph/synthetic follow-ons remain justified
```

## Stop / kill criteria

Stop this experiment if any of the following holds:

- no lawful/reproducible checkpoint can be obtained;
- representation extraction cannot be made deterministic enough for a bounded comparison;
- the frozen representation fails to improve any relevant held-out task;
- only a heavily tuned/full-fine-tuned model appears promising, violating this pilot's cheap-falsification purpose.

Do not respond to a weak result by expanding compute or adding many tasks.

## Non-goals

- no production default changes;
- no full M2BERT fine-tuning;
- no LoRA/PEFT yet;
- no generic symbolic-model service/framework;
- no vector database;
- no new user-facing analysis surface;
- no claim that M2BERT is universal across genres/repertoires.

## Acceptance

- [ ] exact upstream assets/rights status recorded;
- [ ] fixed data/splits frozen before final result-bearing run;
- [ ] >=2 LC-relevant tasks evaluated;
- [ ] baseline, linear probe and tiny MLP compared on the same held-out data;
- [ ] compute remains within the bounded research budget or the issue stops with a reason;
- [ ] result is reproducible from a documented command/environment;
- [ ] explicit FAIL/NARROW/CONTINUE disposition written back to #1303;
- [ ] no production/model-authority change is made from exploratory evidence alone.

Refs #1303 #1297 #1246 #1233 #1300 #1178.

**Comment by gr-rr** [comment]
## Kickoff / pre-registration refinement

Start this as the first result-bearing pilot under #1303. No open PR currently owns #1304.

### Initial tasks

Subject to exact upstream/data rights verification, use SMCBenchmark public labels and start with:

1. **ME — Melody Extraction**: directly LC-relevant note membership task.
2. **MNID — motif-note identification**: second LC-relevant note task and reusable label set for #1305.

Do not add a third task until these two run cleanly and the primary comparison is interpretable.

### Primary comparison

Freeze before full run:

```text
A. task-native/simple literal baseline
B. frozen M2BERT -> linear/logistic head
C. frozen M2BERT -> one tiny MLP
```

Current/specialist LC methods may be reported separately as product ceilings/controls where semantics align, but should not make the frozen-representation question impossible to interpret.

### Embedding extraction constraint

Do **not** use the upstream all-hidden-state pickle format as the durable corpus cache. The upstream extractor emits all 13 states for `(13, 1024, 768)`, which is unnecessarily large for this pilot.

Primary run:

```text
one predeclared hidden layer (start with final layer)
FP16 cache
content-addressed by source/split + checkpoint/revision + extractor revision
actual note count only in durable output where practical
```

An optional small fixed layer ablation is allowed only after the primary result, not as a sweep.

### Runtime ladder

1. local CPU: one-piece extraction smoke + deterministic note/label identity proof;
2. #1311 Kaggle T4: full frozen embedding materialization if materially faster/easier;
3. tiny heads can return to CPU/sklearn/PyTorch after cache exists.

Record wall time, peak RAM if practical, cache bytes/note, accelerator, and full extraction count. These measurements should replace future compute guesses.

### Stop before compute

Before any result-bearing run, explicitly record:
- exact M2BERT source revision;
- checkpoint SHA/identity and rights status;
- SMCBenchmark processed-data revision/source terms for ME and MNID;
- exact train/dev/test or official CV split;
- deterministic note identity mapping.

If the checkpoint/data rights or note mapping cannot be made reproducible, stop rather than spending GPU quota.

Refs #1303 #1305 #1311 #1233 #1300.

**Comment by gr-rr** [comment]
Execution note after Kaggle GPU proof: branch `research/1304-m2bert-probe` is the active owner. Run this as staged materialization rather than one fragile monolith:

1. tiny fixed-sample smoke through exact M2BERT revision/checkpoint;
2. benchmark manifest for ME + MNID frozen before result run;
3. per-piece/shard embedding extraction, only predeclared hidden layer(s), FP16;
4. publish successful reusable embedding cache as a private versioned Kaggle Dataset (or consume directly as a Kaggle kernel source for short-lived staging) so downstream head runs do not recompute the encoder;
5. train/evaluate simple baseline, frozen-embedding logistic/linear head, tiny MLP; heads are cheap enough to rerun from scratch;
6. collect result JSON/plots/runtime and write FAIL/NARROW/CONTINUE back here + #1303.

Kaggle notebook T4 x2 has now been proven end-to-end. The earlier CPU-only behavior was specific to script kernels; the reusable runner is being changed to emit notebooks. Do not add a custom checkpoint server, MLflow service, or sweep framework for this pilot.

**Comment by gr-rr** [comment]
Execution update — first result-bearing frozen probe completed on free Kaggle T4 compute.

Pinned setup:
- M2BERT revision `e296e7f4808b90a619eda81394886cf8c2f8101c`
- checkpoint revision `80efba209265e706756f345af4452fc58b993e49`, SHA-256 `cf15015f8ed1bdf94983812ffb3229e51548b676dc5ceda76b91a09056b40f63`
- processed note-data SHA-256 `faf71e05f312cd3f22c0e67498191e0d07ef1a1c5999b26371eab69f61bd89e3`
- frozen layer 12; 20k train notes, 10k valid, 20k test
- simple control = one-hot CP event fields + sklearn logistic regression
- learned heads = frozen M2BERT + sklearn logistic / 64-unit MLP

Pilot results:

| task | simple control | frozen M2BERT linear | frozen M2BERT tiny MLP |
|---|---:|---:|---:|
| Melody, official test macro-F1 | 0.6241 | **0.9103** | 0.9089–0.9094 (3 seeds) |
| MNID, official CV fold-0 test F1 | 0.3121 | **0.6442** | **0.6526–0.6636** (3 seeds) |

Accuracy: Melody 0.777 control vs 0.946 linear; MNID 0.681 control vs 0.773 linear / 0.775–0.779 MLP.

Runtime was small: Melody extraction train/valid/test ~111 s total and head fitting ~36 s; MNID extraction ~15 s and head fitting ~15 s. Account quota after smoke + both pilots: 0.15 / 30 GPU-hours used.

Interpretation: **strong preliminary CONTINUE** for the representation hypothesis. The lift is large on two materially different note-level tasks, and Melody in particular is almost linearly separable from a frozen representation. Tiny MLP does not improve Melody and only modestly improves MNID, which strengthens the shared-representation interpretation rather than arguing for a large task head.

Do not close yet / do not promote to product yet. Two methodological cleanups are required before final disposition:
1. sklearn reports convergence warnings for the unscaled M2BERT logistic head; rerun the head with standard scaling / a converged linear solver.
2. MNID pilot used only official fold 0; run all 5 official folds (same bounded 20k-note-per-fold budget) or enough folds to establish that fold-0 lift is not anomalous.

Production-rights gate remains unresolved: this is research evidence, not yet a production dependency decision.

**Comment by gr-rr** [comment]
## Current handoff — confirmatory tail only

This issue has already crossed the bar from feasibility smoke to a strong preliminary representation result. Do **not** restart it or add more tasks/models.

Pinned pilot result:

| task | simple control | frozen M2BERT linear | frozen M2BERT tiny MLP |
|---|---:|---:|---:|
| Melody official test macro-F1 | 0.6241 | **0.9103** | 0.9089–0.9094 |
| MNID official fold-0 test F1 | 0.3121 | **0.6442** | **0.6526–0.6636** |

Pinned identities are recorded in the earlier result comment, including M2BERT/checkpoint/data hashes. Total Kaggle usage through smoke + both pilots was only ~0.15 GPU-hours.

Interpretation under the revised #1303 doctrine:

> This is an **OSS representation/adoption study**, not a mandate to train a ListenCloser encoder. The important possible consequence is that a frozen shared symbolic representation might let us delete several bespoke heuristics or support tiny task heads cheaply.

Exactly two validation tasks remain before final disposition:

1. rerun the linear probe with standardization / a solver that demonstrably converges, because the preliminary sklearn linear head emitted convergence warnings;
2. run MNID across all five official CV folds (or the predeclared sufficient multi-fold confirmation) to show fold 0 is not anomalous.

Then write one final `FAIL | NARROW | CONTINUE` decision to this issue and #1303. If `CONTINUE`, do **not** directly productionize M2BERT here: open/route a separate focused adoption owner only after exact code/checkpoint rights, runtime/maintenance, and a concrete legacy-deletion/product consumer are known.

No layer sweep, no fine-tuning, no third task, no generic embedding service.
